CMSPlugin refactoring and magic late initialisation

nikosdion opened the pull request:
Pull Request for Issue #39077 .

### Summary of Changes

* Refactored legacy code out of CMSPlugin and into traits. 6.0 can remove the `use` statements for the legacy implementation traits and 7.0 can remove the traits.
* Added magic late initialisation. See below for more information.

### Testing Instructions

* Install a Joomla 4.3 (dev) site
* Install a bunch of third party extensions if you want

### Actual result BEFORE applying this Pull Request

The site works.

### Expected result AFTER applying this Pull Request

The site still works!

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org:
- [X] No documentation changes for manual.joomla.org needed

### Magic Late Initialisation in CMSPlugin

In the past, initialising the plugin was typically done in its constructor. However, plugin objects are constructed for all published plugins as soon as the plugin group they belong in is imported via `\Joomla\CMS\Plugin\PluginHelper::importPlugin`. Even if the plugin's event handlers are not used the initialisation has executed, with an adverse impact to the site's performance.

This PR introduces a new method, `doInitialise`. Developers can override it and put their plugin's initialisation code in there. This method is called automatically and exactly once, right before the first event handler of this plugin executes. This makes the plugin's initialisation code execute only when it's _truly_ going to be needed during the request's lifetime. If the plugin's event handlers are not used the initialisation code never runs, therefore mitigating the performance impact.

It is possible that a developer might need to disable this feature completely, e.g. because the initialisation is already executed late in the constructor of services which are accessed through an object implementing the Factory pattern as needed — meaning that the initialisation code runs _even later_ than what the `doInitialise` method itself can do. In this case, the plugin can set the `allowLateInitialisation` to `false`.

### Performance impact

I was skeptical as to what the performance impact of late initialisation would be since we are essentially doubling the event handers for each event. To quantify the impact I used Apache Benchmark against the Joomla sample data and the URL `index.php/sample-layouts/blog` which calls as many events as you are going to get with the sample data.

The command line used was `ab -n 500 -c 5 'https://jdev4.local.web/index.php/sample-layouts/blog'`. The site runs on a local server on my MacBook Pro with an Apple M2 processor, 24GiB or RAM, Apache 2.4.54, MySQL 8.0.30, and PHP 8.1.10 (all installed with HomeBrew on macOS Ventura). PHP has OPcache enabled with a revalidation frequency of 2 seconds.

Results **WITHOUT** late initialisation:
```
Time taken for tests:   7.199 seconds

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        2    3   1.0      3       8
Processing:    55   67  15.1     63     176
Waiting:       54   66  14.9     62     175
Total:         57   70  15.2     66     184

Percentage of the requests served within a certain time (ms)
  50%     66
  66%     68
  75%     69
  80%     70
  90%     79
  95%     93
  98%    131
  99%    136
 100%    184 (longest request)
```

Results **WITH** late initialisation:
```
Time taken for tests:   7.224 seconds

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        2    4   1.5      4       9
Processing:    53   66  16.0     62     169
Waiting:       52   65  15.8     61     164
Total:         56   70  16.0     66     175

Percentage of the requests served within a certain time (ms)
  50%     66
  66%     68
  75%     69
  80%     70
  90%     78
  95%     97
  98%    139
  99%    150
 100%    175 (longest request)
```

Conclusion: the performance impact is below what can be reasonable measured. The random performance deviation of the server because of OS load, filesystem access, thermal management, PHP compilation, etc has a far more pronounce impact on the test results than the late initialisation feature.

### Backwards Compatibility Notes

The PR is written in such a way as to maintain full backwards compatibility with Joomla 4.2 (which is fully b/c to Joomla 4.0 and mostly b/c to Joomla 3.10). If a plugins works on Joomla 4.0–4.2 it will still work with this PR applied.

### Deprecation Notes

Here is the intention of this PR with regards to deprecations and b/c management over the next three major versions of Joomla (5, 6, and 7).

Plugins should start using the SubscriberInterface and handle events instead of having callback listeners with a variable number of arguments. The CMSPlugin will stop using the legacy code traits in Joomla 6.0 but the traits will not be removed until Joomla 7.0. This gives a minimum of four and a maximum of six years after Joomla 5.0's release for plugins to be refactored. Note that this is in addition to the five years between the introduction of SubscriberInterface in the Joomla 4 alpha versions and the release of Joomla 5.0.

Plugins should move their initialisation to `doInitialise`. The constructor **WILL NOT** be made final as it still needs to be overridden when the developer wants to push services to the plugin object through the constructor in the service provider. Any initialisation in the constructor beyond assigning services pushed by the service provider is, however, now considered a bad practice and not in line with the philosophy of the Joomla architecture.

Plugins should not expect that the `loadLanguage` method will be available by default. Plugins should use the `\Joomla\CMS\Plugin\LanguageAwareTrait` to make sure the `loadLanguage` method is available. The CMSPlugin class will stop loading this trait automatically in Joomla 6.0. Using this trait until Joomla 6.0 is released **WILL NOT** cause any adverse affects to plugins. Therefore, plugin developers can start refactoring their code as soon as this PR is merged.

The `autoloadLanguage` property is now considered deprecated and will be removed in Joomla 6.0. Plugins need to use the `\Joomla\CMS\Plugin\LanguageAwareTrait` and call `$this->loadLanguage()` manually in their `doInitialise` method instead of setting `autoloadLanguage` to true. **DO NOT** try to load any language files in the constructor.

Commit:
CMSPlugin refactoring

Move application object management into a new ApplicationAwareTrait

Commit:
CMSPlugin refactoring

Move legacy and optional stuff into traits

Commit:
Magic late initialisation in CMSPlugin

Commit:
CMSPlugin refactoring

I forgot to add something here.

richard67 commented:
@nikosdion The issue can be closed as we have this PR now? Or is there something remaining so it should be kept open?

nikosdion commented:
@richard67 Yes, you can close #39088 as this PR fully addresses the reported problem (and then some).

richard67 commented:
@nikosdion Well, since you are the author of the issue, you should have the privilege to close it yourself. It would have saved us time.

richard67 commented:
@nikosdion Unit tests for the CMS plugin are failing, see https://ci.joomla.org/joomla/joomla-cms/59133/1/11 . Possibly they need some update, too?

brianteeman commented on libraries/src/Plugin/CMSPlugin.php:
```diff
@@ -47,58 +51,62 @@ abstract class CMSPlugin implements DispatcherAwareInterface, PluginInterface
-     * Constructor
+     * Constructor.
+     *
+     * Do not put any slow initialisation code in it, e.g. code which accesses the database,
+     * performs lengthy calculations, or calls external services over HTTP. Put this code into the
+     * doInitialise() method which is called ONCE, before the first event handler in your plugin is
+     * executed.
```
```suggestion
     * Do not put any slow initialisation code in the constructor, e.g. code which accesses the database,
     * performs lengthy calculations, or calls external services over HTTP. Put that code into the
     * doInitialise() method which is called ONCE, before the first event handler in your plugin is
     * executed.
```

clarifies what "it" and "this code"

brianteeman commented on libraries/src/Plugin/CMSPlugin.php:
```diff
@@ -109,267 +117,113 @@ public function __construct(&$subject, $config = array())
+     * This method is "magically" called exactly once, right before the very first time an event
+     * handler in your plugin is called. This makes sure that all lengthy initialisation code in
+     * your plugin will only be executed if your plugin is used in a page load, drastically
+     * improving the site's performance on page's where your plugin is loaded but its event handlers
```
```suggestion
     * improving the site's performance on pages where your plugin is loaded but its event handlers
```

grammar

brianteeman commented on libraries/src/Plugin/LanguageAwareTrait.php:
```diff
@@ -0,0 +1,36 @@
+<?php
```
copyright headers ?

nikosdion replied:
Thanks! Just fixed.

brianteeman commented on libraries/src/Plugin/LegacyListenerTrait.php:
```diff
@@ -0,0 +1,221 @@
+            $param = array_shift($parameters);
+            $paramName = $param->getName();
+
+            // No type hint / type hint class not an event or parameter name is not "event"? It's a legacy listener.
```
something doesn't make sense here. is there a typo or a missing word?

nikosdion replied:
I just moved this code from CMSPlugin into the trait — keeping the comments as-is.

It does make sense if you understand the context of the code. If there is no type hint on the parameter OR the type hint of the parameter is a class other than Event OR the parameter's name is not event THEN it's a legacy listener.

brianteeman replied:
i read it about 5 times - i'm sure there is missing something - perhaps just punctuation. anyway your call.

Commit:
Update libraries/src/Plugin/CMSPlugin.php

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update libraries/src/Plugin/CMSPlugin.php

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Missing copyright header

wilsonge commented on libraries/src/Plugin/PluginHelper.php:
```diff
@@ -228,7 +229,19 @@ protected static function import($plugin, $autocreate = true, DispatcherInterfac
+         *
+         * @deprecated 5.0 Use SubscriberInterface in plugin objects instead
+         */
+        if (method_exists($plugin, 'registerListeners')) {
```
https://github.com/joomla/joomla-cms/blob/4.2-dev/libraries/src/Extension/PluginInterface.php probably better?

nikosdion replied:
No.

The PluginInterface must NOT have registerListeners. This is a legacy method which we are deprecating in 5.0. In 6.0 the CMSPlugin will no longer use by default the trait which implements it. In 7.0 the trait will be dropped and so will the code here.

wilsonge replied:
Well that method is the only method in the interface and is also what CMSPlugin currently implements (if I'm honest given the current structure I'm not sure why - but that's the implementation we have) so surely it makes sense to make advantage of it given it's in the code)

nikosdion replied:
I will put initialisePlugin in the interface and mark the registerListeners as deprecated. We really need to move away from registerListeners and have everyone use SubscriberInterface. That's the only way to stop using reflection on plugins which has a much more significant impact on the site's performance than pretty much anything else we might do in the core.

laoneo replied:
I think as long as we can't have the concrete events with listeners like `onContentPrepare ($context, $item)` we will never be able to remove registerListeners, or do I miss here something?

nikosdion replied:
You are wrong :)

Since you are using Akeeba Release System on your site I think you will recognise this plugin: https://github.com/akeeba/release-system/blob/development/plugins/content/arslatest/src/Extension/Arslatest.php#L96

As you can see, it is doing
```php
public function onContentPrepare(Event $event)
```
and it uses SubscriberInterface just fine.

I would never tell other developers to do something I have not already figured out how to do. I need to be able to teach them how to do it before I can even think about asking them to do it. Fair?

nikosdion replied:
PS: To be fair, I have a bit of an advantage using Event in plugins as I had contributed the first pass of that implementation in Joomla 4 back in August 2015. I think it was actually the first code we had written in the J4 Working Group. I was aching to use that oh-so-much-better approach to writing plugins ever since. It only took 6 years to get there! 🤣

laoneo replied:
But I was talking about the old signature: `onContentPrepare ($context, $item)`

nikosdion replied:
The old signature requires the legacy code.

However, you **do not** need concrete events to changes the old signature to a signature using Event. You only need to make _small_ changes in the implementation code.

Therefore it is absolutely practical to convert all of the core plugins to the new signature before we complete the work on concrete events.

When we complete the work on concrete events we can revisit the core plugins and change the signature to use the concrete event class instead of the generic Event class.

In fact, I think we should do that **within the 4.x lifecycle** (so, in 4.3) to make sure we don't have any edge cases which would screw us later on. I already have qualms about what to do with fields plugins (I actually meant to ask you about them) and I know for sure that `editor-xtd` plugins are NOT plugins, they are instantiated directly by the editor and have their onDisplay method (which is NOT an event handler!) called directly. IMHO we should change that architecture in the editor but I am not as familiar with it to propose changes — I am unsure about the b/c impact. We'd have to ask the 3 or 4 people who have written alternative editors for Joomla for their input.

laoneo replied:
> The old signature requires the legacy code.

That was what I meant, and this will stay for a long time.

nikosdion replied:
But it does NOT have to stay, this is what **I** am saying!

Look at my plugin. I took the onContentPrepare handler from the Joomla 3 plugin and wrote it Joomla 4 native.

Why would Joomla itself insist on _idiotically_ keeping the old code when it can be rewritten to be Joomla 4 native?! It makes no sense.

laoneo replied:
It has to stay because of 3rd party extensions, otherwise you will kill all of them which are written the old way. And it would be impossible to be able to run the same plugin on 3 and upwards. Anyway, just wanted to clarify that removing that registerListeners function comes hand in hand with dropping support for the old signature. I'm resolving this conversation, to not loose focus now.

nikosdion replied:
Dude, you are not killing any extension. Joomla 4 does not use legacy listeners at all. It only uses event listeners.

Have you noticed that registerListeners converts legacy listeners to event listeners? It just passes all the arguments, in the order defined, to the callback and converts the callbacks return value into the result argument of the event.

The whole point I did that 7 years ago was to NOT break sites when they have a mix of old and new listeners.

If you look at my PR for concrete events this is something I had explicitly talked about. See the massive code comment in \Joomla\CMS\Event\ReshapeArgumentsAware where I have already documented this!

wilsonge commented on libraries/src/Plugin/AutoloadLanguageAware.php:
```diff
@@ -0,0 +1,14 @@
+
+namespace Joomla\CMS\Plugin;
+
+interface AutoloadLanguageAware
```
What's the aim of this interface? Seems unused

nikosdion replied:
It is unused, I am removing it. 

It was part of a test I was making, trying to have language autoloading by implementing an interface. It was not a very good fit for making CMSPlugin lighter. I dropped the code and forgot to remove the interface.

richard67 commented:
Unit tests still failing, see my previous comment.

Commit:
Remove unused interface

Commit:
Clarify badly phrased comment

This fixes a comment copied over from
the existing codebase, introduced in
Joomla 4.0.

Commit:
Correctly register late init for legacy events

Commit:
Tests for early initialisation

Commit:
s/early/late

Clearly, I need more coffee

nikosdion commented:
@richard67 I was asleep last night :) I have fixed the tests today and added more tests for the late initialisation code.

richard67 commented on tests/Unit/Libraries/Cms/Plugin/CMSPluginTest.php:
```diff
@@ -249,14 +251,16 @@ public function testNotLoadLanguageWhenExists()
      * @return  void
      *
-     * @since   4.2.0
+     * @since   __DEPLOY_VERSION__
```
```suggestion
     * @since   4.2.0
```
I don't think we should change the since tag here.

richard67 commented:
> @richard67 I was asleep last night :) I have fixed the tests today and added more tests for the late initialisation code.

@nikosdion Sure, you've deserved to sleep and eat and live of course. I did not want to push, only wanted to be sure it doesn't get lost within all the other review comments. Thanks for fixing the unit tests. It will make the other tests (API and system) run, too.

Commit:
Update tests/Unit/Libraries/Cms/Plugin/CMSPluginTest.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

Commit:
Update the PluginInterface.

Commit:
Update tests

We need to set allowLateInitialisation
to false for CMSPlugin to only
register one listener per event.

Commit:
Merge branch '4.3-dev' into feature/39077_cmsplugin_refactoring

nikosdion commented:
I am not going to be contributing to this project anymore. I won't be replying to any questions or making any changes until then. I will close any outstanding PRs and delete their code on December 31st, 2022. Merge whatever you want until then.

nikosdion commented:
I guess plugins still going through the CMSFactory to load their language is not considered a problem, given it's three months with no sign of anyone willing to merge it, so I am closing this PR.
